runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -597,10 +607,58 @@ Table<K, V, M> openTable(@Nonnull final String namespace,
                 mapSupplier);
         tableMap.put(fullyQualifiedTableName, (Table<Message, Message, Message>) table);
 
+        resetIfTableTrimmed(namespace, tableName);
         registerTable(namespace, tableName, kClass, vClass, mClass, tableOptions);
         return table;
     }
 
+    /**
+     * If the table that is going to be registered, is not there in the registry table,
+     * but the address space seems to contain the stream map with no content, it means
+     * the table was trimmed without being checkpointing. If a trimmed table of this sort
+     * is found, append a HOLE to the stream to reset it.
+     *
+     * @param namespace Namespace of the table
+     * @param tableName Name of the table
+     */
+    private void resetIfTableTrimmed(String namespace, String tableName) {
+        TableName tableNameKey = TableName.newBuilder()
+                .setNamespace(namespace)
+                .setTableName(tableName)
+                .build();
+        CorfuRecord<TableDescriptors, TableMetadata> corfuRecord = this.registryTable.get(tableNameKey);
+        UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(namespace, tableName));
+        StreamAddressSpace streamAddressSpace = this.runtime.getSequencerView()
+                .getStreamAddressSpace(new StreamAddressRange(streamId, Address.MAX, Address.NON_ADDRESS));
+        if (streamAddressSpace.size() == 0 && streamAddressSpace.getTrimMark() != Address.NON_ADDRESS && corfuRecord == null) {
+            log.info("Found trimmed table that is re-opened. Reset table {}", getFullyQualifiedTableName(namespace, tableName));
+            resetTrimmedTable(namespace, tableName);
```

Comment:
This looks prone to race conditions. This will be executed by all JVMs on table open, right?

For example, consider the scenario where two threads, T1 and T2, from two different JVMs enter this code path when the if-condition is true. T1 goes first and performs the holefill and an empty checkpoint. Now before T2 has a chance to perform the same, T1 performs new writes on this table. Now, when T2 resumes and performs the above, it seems like these updates can be lost for some versions since the address of the hole made by T2 will be greater than the new updates made by T1.

Reply:
Appending empty checkpoints should not cause issues with writes since the checkpoint is only ever resolved from when there is a trimmed exception hit on the main stream. so in your example T2 appending a checkpoint to the stream should not cause any issues with T1 writing to the same stream right?
However, I think it might be arguably safer to do the registry table check and the checkpoint append in a single transaction if possible so they are atomic.

Reply:
I agree with Zach. If T1 has finished writing HOLE, empty checkpoint and some data writes, then T2 runs L651 (`snapshot = cpw.forceNoOpEntry()`) which obtains a snapshot that's greater than T1's data writes, then T2 will wipe out T1's writes and escape from the conflict detection.

Reply:
Having read registry table and empty checkpointing in the same transaction is not enough because writing HOLE and empty checkpoints do not trigger a conflict. One possible solution is to delete the entry from RegistryTable in the same transaction.

Reply:
I agree with Zach too. Will discuss in slack about updated solution.

Reply:
It is not clear how writing an empty checkpoint to a `checkpoint stream` can overwrite data written in the main stream. Can you please clarify? This sort of a pattern happens all the time today with the compactor right - the compactor appends writes to the checkpoint stream while simultaneously there are writes going into the main stream.

Btw, can we just move this logic to the transaction that also adds the entry in the table registry L275? If there are multiple threads trying to insert the same record into the table on re-open, the worst that can happen is that since the checkpoint table write may not cause a conflict, the two threads append empty checkpoints to the trimmed table which is also supported today since we can have potentially multiple checkpoints for a table right?
But when one of them retries their aborted transaction, they will find that there is no need to reset the trimmed table.

Reply:
`the compactor appends writes to the checkpoint stream while simultaneously there are writes going into the main stream.` Compactor checkpoint needs to use a snapshot and write the coalesced state of the table at that snapshot. During address discovery, if the streams has not been initialized, the reader will first get a highest checkpoint whose snapshot is X, and assumes all updates prior to address X has been covered by this checkpoint. However, in this PR, it starts a snapshot and appends empty state even if there is already data written to the table after reset and reopen. Those writes are valid writes, but they are discarded.

Relevant code: https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java#L447

Reply:
Right, but when we take the snapshot of the transaction that is adding the entry into the table registry in L273 or L280 (new code), and use that to write the empty checkpoint, then automatically if there are writes made which are newer than that (T1's writes in Zach's example), then the object layer will be able to realize that the writes made to the main stream are newer and will not be impacted by the older snapshot checkpoint stream write (T2 in the example).  Please correct me if I am wrong?

Reply:
If another client that hasn't performed a read on the table yet, then they will see the newer checkpoint that does not capture the writes made by T1.

Reply:
how will a client read a table before opening it first?

Reply:
Sorry about the confusion. That's not what I'm saying. A table has to be opened before a read is performed, but opening a table does not trigger a sync.

Reply:
Let's say T1 resets the stream, appends an empty checkpoint to it, reopen the table and write key1 at address 10. A parallel thread resets the stream and appends an empty checkpoint whose snapshot is 15. Now that if there is T3 who just bootstrapped, opened the table and trying to read the table. It will read the CP@15 and assumes any updates to the stream before address 15 has been covered by the CP, including the key1 written at 10.

Reply:
So here is the workflow:
T1: openTable() -> 
        start TX, 
           if record is not in registry table -> if table is trimmed, 
                             use current tx's snapshot,
                             append empty checkpoint,
                             Add registry table entry
             END TX -> say this is address 10
T1 -> writes to table just opened, but that address will be _higher_ than 10, say 11

Now if another thread T2 does the same flow, at the same time, the only way that it can also append an empty checkpoint is if it started when the stream tail was BEFORE 10 (say 9), and when that transaction tries to commit,
it's openTable() will abort and it will retry, when it retries, it will _not_ append an empty checkpoint on retry because it finds that the table registry already has an entry.
Now if it reads data, here is what it will see for that table
Checkpoint stream: Address 10, empty, address 10 empty checkpoint
Main stream: Address 11 (write)

Now main stream's write at address 11 is higher than empty checkpoint whose snapshot is at address 10, so why will write at 11 be lost?

Please let me know if I am still missing something or if an offline discussion might help?

Reply:
I think what you described is correct. The parallel transaction will be aborted due to conflict on updating the same key in RegistryTable. Zach and I are talking about the concurrent execution of `resetIfTableTrimmed()` alone, before Sravanthi added `resetIfTableTrimmed()` to the `openTable()`.